doc/stdenv/platform-notes.chapter.md

```diff
@@ -46,22 +46,189 @@ Some common issues when packaging software for Darwin:
+
+  - If a package supports multiple SDK versions, use the lowest supported SDK version by the package (but no lower than the default SDK). That ensures maximal platform compatibility for the package.
+
+  - If a package specifies a range of supported SDK versions _and_ a minimum supported version, assume the package is using availability checks to support the indicated minimum version.Add latest supported SDK package and a `darwinMinVersionHook` set to the minimum supported by the upstream package.
+
+  Warning: Avoid using newer SDKs than an upstream package supports. When a binary is linked on Darwin, the SDK version used to build it is recorded in the binary. Runtime behavior can vary based on the SDK version, which may work fine but can also result in unexpected behavior or crashes when building with an unsupported SDK.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "foo-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_15 (darwinMinVersionHook "10.15") ]; # Upstream builds with the 15.0 SDK but supports 10.15.
+  }
+  ```
+
+- Libraries that require a minimum version can propagate an appropriate SDK and `darwinMinVersionHook`. Derivations using that library will automatically use an appropriate SDK and minimum version. Even if the library builds with a newer SDK, it should propagate the minimum supported SDK. Derivations that need a newer SDK can add it to their `buildInputs`.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "libfoo-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_15 ]; # Upstream builds with the 15.0 SDK but supports 10.15.
+    propagatedBuildInputs = [ apple-sdk_10_15 (darwinMinVersionHook "10.15") ];
+  }
+  # ...
+  stdenv.mkDerivation {
+    name = "bar-1.2.3";
+    # ...
+    buildInputs = [ libfoo ]; # Builds with the 10.15 SDK
+  }
+  # ...
+  stdenv.mkDerivation {
+    name = "baz-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_12 libfoo ]; # Builds with the 12.3 SDK
+  }
+  ```
+
+- Many of the SDK libraries are built from source (typically corresponding the source releases for the latest release of macOS). Several of these are propagated to your package automatically. They can be accessed via the `darwin` package set.
+
+  - libiconv
+  - libresolv
+  - libsbuf
+
+  Other common libraries are available in Darwin-specific versions with modifications from Apple. Note that these packages may be made the default on Darwin in the future.
+
+  - ICU (compatible with the top-level icu package, but it also provides `libicucore.B.dylib` with an ABI compatible with the Darwin system version)
+  - libpcap (compatible with the top-level libpcap, but it includes Darwin-specific extensions)
+
+- There are two legacy patterns that are being phased out. These patterns were used in the past to change the SDK version. They have been reimplemented to use the `apple-sdk` packages.
```

Comment:
Perhaps call out the gotcha more explicitly - that while the names are similar, they are also different (underscores vs dashes). When I started looking at the new SDK, I was confused by this for a bit.

Reply:
I added a section on the legacy SDKs and their frameworks along with guidance on how to migrate if you were depending on internal details (such as the filesystem location of frameworks).